bottlerocket/agents/src/userdata.rs

```diff
@@ -0,0 +1,106 @@
+use resource_agent::provider::{IntoProviderError, ProviderResult, Resources};
+use toml::{map::Entry, Value};
+
+/// This modifies the first given toml Value by inserting any values from the second Value.
+///
+/// This is done recursively.  Any time a scalar or array is seen, the left side is set to match
+/// the right side.  Any time a table is seen, we iterate through the keys of the tables; if the
+/// left side does not have the key from the right side, it's inserted, otherwise we recursively
+/// merge the values in each table for that key.
+///
+/// If at any point in the recursion the data types of the two values does not match, we error.
+pub fn merge_values<'a>(merge_into: &'a mut Value, merge_from: &'a Value) -> ProviderResult<()> {
+    // If the types of left and right don't match, we have inconsistent models, and shouldn't try
+    // to merge them.
+    if !merge_into.same_type(merge_from) {
+        IntoProviderError::context(
+            None,
+            Resources::Clear,
+            "Cannot merge mismatched data types in given TOML",
+        )?
+    }
+
+    match merge_from {
+        // If we see a scalar, we replace the left with the right.  We treat arrays like scalars so
+        // behavior is clear - no question about whether we're appending right onto left, etc.
+        Value::String(_)
+        | Value::Integer(_)
+        | Value::Float(_)
+        | Value::Boolean(_)
+        | Value::Datetime(_)
+        | Value::Array(_) => *merge_into = merge_from.clone(),
+
+        // If we see a table, we recursively merge each key.
+        Value::Table(t2) => {
+            // We know the other side is a table because of the `ensure` above.
+            let t1 = merge_into.as_table_mut().unwrap();
```

Comment:
Easier to read if we know which one is from/to. 

```suggestion
        Value::Table(from) => {
            // We know the other side is a table because of the `ensure` above.
            let to = merge_into.as_table_mut().unwrap();
```